lib/cambiatus/upload.ex

```
@@ -1,6 +1,10 @@
 defmodule Cambiatus.Upload do
   @moduledoc "Handles file uploading"
 
+  @doc """
+  Saves a file on the configured Amazon S3 bucket
+  """
+  @spec save_file(String.t(), String.t()) :: {:ok, String.t()} | :error
```

Good! Typing functions shows important stuff! You have two different types on this function. In Elm, you can't even do this, you would have to introduce a type that can have two different values. You'll see that in elixir we usually have {:ok, ""} and {:error, ""} as the most common response types. This way you will not hide the error, and give a chance for the consumer of this function to do different things depending on the error.

You did this pattern of having two different types in a lot of functions on this module, please let me know your thoughts on this, maybe I'm not seeing something here!

---

That's fair enough. In cases where there's no relevant information to add on the success type, do we just return `{:ok, ""}` with an empty string?

---

No, usually I do return something that is not an empty string. In the case of the functions you did, you could just return boolean values since the error messages are always the same. I'll comment on them.